core/src/test/java/nu/yona/server/subscriptions/service/migration/EncryptFirstAndLastNameTest.java

```diff
@@ -0,0 +1,166 @@
+{
+	private static final String PASSWORD = "password";
+	private static final Field firstNameUserField = JUnitUtil.getAccessibleField(User.class, "firstName");
+	private static final Field lastNameUserField = JUnitUtil.getAccessibleField(User.class, "lastName");
+	private static final Field firstNameUserPrivateField = JUnitUtil.getAccessibleField(UserPrivate.class, "firstName");
+	private static final Field lastNameUserPrivateField = JUnitUtil.getAccessibleField(UserPrivate.class, "lastName");
+	private static final Field userPrivateField = JUnitUtil.getAccessibleField(User.class, "userPrivate");
+	private User richard;
+
+	@Autowired
+	private MigrationStep migrationStep;
+
+	@Rule
+	public MethodRule cryptoSession = new CryptoSessionRule(PASSWORD);
+
+	@Autowired
+	private BuddyAnonymizedRepository buddyAnonymizedRepository;
+
+	@MockBean
+	private MessageService mockMessageService;
+
+	@Captor
+	private ArgumentCaptor<Supplier<Message>> messageSupplierCaptor;
+
+	@Before
+	public void setUpPerTest() throws Exception
+	{
+		try (CryptoSession cryptoSession = CryptoSession.start(PASSWORD))
+		{
+			richard = JUnitUtil.createRichard();
+		}
+	}
+
+	@Override
+	protected Map<Class<?>, Repository<?, ?>> getRepositories()
+	{
+		Map<Class<?>, Repository<?, ?>> repositoriesMap = new HashMap<>();
+		repositoriesMap.put(BuddyAnonymized.class, buddyAnonymizedRepository);
+		return repositoriesMap;
+	}
+
+	@Test
+	public void upgrade_moveFirstAndLastName_firstAndLastNameMoved() throws IllegalArgumentException, IllegalAccessException
+	{
+		// First and last name are by default not set on the User entity anymore, so set them explicitly
+		String firstName = "First name";
+		String lastName = "Last name";
+		firstNameUserField.set(richard, firstName);
+		lastNameUserField.set(richard, lastName);
+
+		migrationStep.upgrade(richard);
+
+		assertThat(firstNameUserPrivateField.get(userPrivateField.get(richard)), equalTo(firstName));
+		assertThat(lastNameUserPrivateField.get(userPrivateField.get(richard)), equalTo(lastName));
+		assertThat(firstNameUserField.get(richard), is(nullValue()));
+		assertThat(lastNameUserField.get(richard), is(nullValue()));
+
+		verify(mockMessageService, times(1)).broadcastMessageToBuddies(Matchers.<UserAnonymizedDto> any(),
+				messageSupplierCaptor.capture());
+		Message message = messageSupplierCaptor.getValue().get();
+		assertThat(message, instanceOf(BuddyInfoChangeMessage.class));
+		BuddyInfoChangeMessage buddyInfoChangeMessage = (BuddyInfoChangeMessage) message;
+		assertThat(buddyInfoChangeMessage.getNewFirstName(), equalTo(firstName));
+		assertThat(buddyInfoChangeMessage.getNewLastName(), equalTo(lastName));
+		assertThat(buddyInfoChangeMessage.getNewNickname(), equalTo(richard.getNickname()));
+		assertThat(buddyInfoChangeMessage.getMessage(), equalTo("User changed personal info"));
+	}
+
+	@Test
+	public void upgrade_moveFirstAndLastNameTwice_idempotent() throws IllegalArgumentException, IllegalAccessException
```

Comment:
How could it be the case that `upgrade` is called twice?

Reply:
It's a safety measure. IMHO, upgrade steps need to be idempotent, just to be sure.